tests/rptest/tests/full_disk_test.py

```diff
@@ -520,3 +520,84 @@ def test_disk_stat_free_delta_overrides(self):
+            # when log storage exceeds 5 MB then kafka produce api is going to
+            # be blocked
+            log_storage_max_size=5 * 1 << 20,
+
+            # the test is intended to show that GC will occur immediately
+            # when a low disk space situation arises. to test this, we want
+            # to avoid any automatic / scheduled GC so we set a housekeeping
+            # interval that ensures that housekeeping won't run a scheduled
+            # job during the test.
+            log_compaction_interval_ms=24 * 60 * 60 * 1000,
+
+            # using a small 0.25 mb log segment size provides "liquidity" to the
+            # test for lack of a better word. we need the system to have reclaim
+            # opportunities and so we want to avoid data being stuck in the
+            # active head segment.
+            log_segment_size=1 << 18,
+
+            # wait until reaching the max size before trying to run GC
+            log_storage_max_size_low_watermark=1.0)
+        super().__init__(test_context=test_ctx, extra_rp_conf=extra_rp_conf)
+
+    @cluster(num_nodes=3)
+    def test(self):
+        kafka_tools = KafkaCliTools(self.redpanda)
+        kafka_tools.produce(self.topic, 40 * 1024, 1024, throughput=2048)
+        assert self.redpanda.search_log_any(
+            "rejecting produce request: no disk space")
+        assert self.redpanda.search_log_any(
+            "Setting maximum size exceeded flag")
+        assert self.redpanda.search_log_any(
+            "Clearing maximum size exceeded flag")
+
+
+class LogStorageLowWatermarkTest(RedpandaTest):
+    """
+    Test that GC runs before reaching the max size at which point produce api
+    woudl be blocked. Very similar to the max size test above, except we reduce
+    the low watermark and write at a slower rate to give the system time to
+    respond before blocking without having to have a "larger" test.
+    """
+    topics = (TopicSpec(partition_count=1,
+                        retention_bytes=1,
+                        retention_ms=1,
+                        cleanup_policy=TopicSpec.CLEANUP_DELETE), )
+
+    def __init__(self, test_ctx):
+        extra_rp_conf = dict(log_storage_max_size=20 * 1 << 20,
+                             log_compaction_interval_ms=24 * 60 * 60 * 1000,
+                             log_segment_size=1 << 18,
+                             log_storage_max_size_low_watermark=0.5)
+        super().__init__(test_context=test_ctx, extra_rp_conf=extra_rp_conf)
+
+    @cluster(num_nodes=3)
+    def test(self):
+        kafka_tools = KafkaCliTools(self.redpanda)
+        kafka_tools.produce(self.topic, 60 * 1024, 1024, throughput=1024)
+        # should not see anything that would indicate we reached the max size
+        assert not self.redpanda.search_log_any(
+            "rejecting produce request: no disk space")
+        assert not self.redpanda.search_log_any(
+            "Setting maximum size exceeded flag")
```

Comment:
I see that the HWM correctly triggers GC dropping usage back down below HWM, but then oddly, storage space continues to increase after the first round of GC, without seeming to trigger additional rounds of GC. I think that the housekeeping loop is getting "stuck" somewhere and not getting back to its high priority GC task. Still investigating that.